### ISSUE TITLE
Cannot suppress Pylance diagnostic errors in Python library files when try to set up configuration options

### ISSUE THREAD

**Issue by ivanzemskiy** [microsoft/pylance-release#3793]
## Environment data

-   Language Server version: 2022.12.20 (pyright 621d886b)
-   OS and version: macOS 10.15.7, Windows 22H2
-   Python version: 3.11 (official distribution from python.org)

## Repro Steps

1. Open empty folder (project root)
2. Create 1 file:
   - ./pyproject.toml
      ```toml
      [tool.pyright]
      typeCheckingMode = "strict"
      ```
3. Navigate to `.py` / `.pyi` file from the Python library folder (tested with `typing.py`) or Pylance's typeshed stubs directory (tested with `builtins.pyi`)

**Note**: I've tried to add the absolute path to my Python library to the next settings: "python.analysis.exclude" / "python.analysis.ignore" from [language server settings](https://code.visualstudio.com/docs/python/settings-reference#_python-language-server-settings), "exclude" / "ignore" fields from [pyright config options](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#main-pyright-config-options). As I understand, "python.analysis.ignore" is that what I need. But it does not work. None of these settings affect the situation. The issue is only fixed when I delete or rename `pyproject.toml` file. But in this case I can't set up [Pyright configuration](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#type-check-diagnostics-settings).

There is the same issue if you try to use `pyrightconfig.json` instead of `pyproject.toml`.

## Expected behavior

I want to see Python library files without diagnostic output (errors and warnings) in "strict"/"basic" type checking mode and have the ability to tune Pylance/Piright configuration.

## Actual behavior

<img width="600" alt="Type checking errors: macOS" src="https://user-images.githubusercontent.com/5605262/209953998-5b6f54b4-93f5-4d1c-9f88-29b7f6e5346a.png">

## Logs
[vscode-macos.txt](https://github.com/microsoft/pylance-release/files/10319862/vscode-macos.txt)

**Comment by erictraut** [comment]
This is the intended behavior. If you open a file that is outside of your project, pylance applies the diagnostic rules that you've selected in your settings. 

The "ignore" and "exclude" options aren't exposed as a setting currently, so if you want to ignore or exclude certain files or directories, you'll need to add those to the pyproject.toml file.

Another option if you want to use strict-mode type checking for all files in your project... you can set typeCheckingMode to "off" and then add a "strict" section to your pyproject.toml and specify a directory of ".", which indicates that every file under your project should use strict type checking.

**Comment by ivanzemskiy** [comment]
Thanks for the idea with `"strict"` section. Yes, the next `pyproject.toml` config works for me:

```toml
[tool.pyright]
typeCheckingMode = "off"
strict = [
    "**/*.py",
]
```

But, if I add some of [diagnostic rules](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-rule-defaults) to that `pyproject.toml` and, even if I add them with their default 'error' values, I’ll start getting the errors in non-project files for these added rules again.

However, it's a problem, in fact, only for a small number of rules (whose values are 'none' or 'warning' in 'strict' mode) - rules, whose security level can be increased.

**Comment by ivanzemskiy** [comment]
I think, I've found the solution for my problem. 

I've found, that I can use `"python.analysis.diagnosticSeverityOverrides"` and, it was quite surprising to me, I can decrease! the security level of specific rules. I totally deleted the "[tool.pyright]" section from my `pyproject.toml` and moved the Pyright configuration to VSCode's project `settings.json` file:

.vscode/settings.json
```json
"python.analysis.typeCheckingMode": "strict",
"python.analysis.diagnosticSeverityOverrides" : {
    "reportUnusedImport": "warning",
    "reportUnusedClass": "warning",
}

```
User-level settings.json
```json
"python.analysis.ignore": [
    "/**/Python.framework",
    "/**/.vscode/extensions",
],
```

Without "[tool.pyright]" section, `"python.analysis.ignore"` began to work again and it started to suppress the errors in library files. That's what I wanted.

**Comment by ivanzemskiy** [comment]
But, my solution cannot be achieved by using only pyproject.toml and you need to specifically add another project config file settings.json. That's not always convenient. I mean, sometimes you want to keep everything in one configuration file.

**Comment by ivanzemskiy** [comment]
I think, I've found the solution for my problem. 

I've found, that I can use `"python.analysis.diagnosticSeverityOverrides"` and, it was quite surprising to me, I can decrease the security level of specific rules. I've totally deleted the "[tool.pyright]" section from my `pyproject.toml` and moved the Pyright configuration to VSCode's project `settings.json` file:

.vscode/settings.json
```json
"python.analysis.typeCheckingMode": "strict",
"python.analysis.diagnosticSeverityOverrides" : {
    "reportUnusedImport": "warning",
    "reportUnusedClass": "warning",
    "reportMissingSuperCall": "error",
}

```
User-level settings.json
```json
"python.analysis.ignore": [
    "/**/Python.framework",
    "/**/.vscode/extensions",
],
```

Without "[tool.pyright]" section, `"python.analysis.ignore"` began to work again and it started to suppress the errors in library files. That's what I wanted.

**Comment by ivanzemskiy** [comment]
But, using my solution you need to specifically add settings.json to your project.

Inconvenience of this approach is that you need to keep you linters configuration in several config files, for e.g. Mypy’s configuration in pyproject.toml and Pylance’s configuration in .vscode/settings.json.

Sometimes there is a need to keep everything in one place.

**Comment by ivanzemskiy** [comment]
I've found the case when my solution from [#issuecomment-1368169035](https://github.com/microsoft/pylance-release/issues/3793#issuecomment-1368169035) does not work. Minimal configuration:

1. Client configuration (macOS)
   - ~/Library/Application Support/Code/User/settings.json
     ```
     clean
     ```
2. Remote server configuration (Windows, connected via [Remote - SSH](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh) extension)
   - ~/.vscode-server/data/Machine/settings.json
     ```json
     "python.analysis.ignore": [
         "c:/**/Python*/Lib",
         "c:/**/Python*/site-packages",
         "c:/**/.vscode-server/extensions",
     ],
     ```
3. Remote workspace
   - .vscode/settings.json
     ```json
     "python.analysis.typeCheckingMode": "strict",
     ```

If I just open the file `~\.vscode-server\extensions\ms-python.vscode-pylance-2023.1.10\dist\typeshed-fallback\stdlib\builtins.pyi`, there are no errors. 

But if I add test.py to my workspace, type `a = list` and navigate to `class list` inside builtins.pyi, then I get:

![test-pylance-remote](https://user-images.githubusercontent.com/5605262/211163636-a35eb320-2c1f-4e16-ace3-a8d2c96ae0fe.jpg)
[pylance-ls_2023_1_10-log.txt](https://github.com/microsoft/pylance-release/files/10366956/pylance-ls_2023_1_10-log.txt)

**Note**: reproduced only on remote, can't reproduce locally

**Comment by heejaechang** [comment]
it looks like when there are multiple places that set settings, Pylance, rather than merging them, one completely replaces the other. let me check whether that is the case and discuss whether we should rather merge rather than replace.

**Comment by samskiter** [comment]
Also having issues with this - my use case is I have a submodule checked out in my repo. I've installed that with pip in development mode. It's a 3p library I'm working with and need to make a couple of alterations - mypy/pylance is peppering it with red. It doesn't do the same for my venv folder.

Interestingly, if I copy the typing_extensions.py file from the root of my venv to the root of the submodule, all is well!

It seems impossible to get the same behaviour for an arbitrary directory as you can for venv... In my case, if pylance were *really* smart, it would be able to detect that my submodule is installed as a pip module and apply the same behaviour to it...

**Comment by heejaechang** [comment]
@samskiter  I dont think it is the same issue as the original issue. can you open a new issue so we can track it separately with more detail such as repro steps we can follow and logs?

thank you
- heejae

**Comment by heejaechang** [comment]
current behavior is by design. for what you want, you should just use `vscode`'s error list `filter` to remove errors from those outside of workspace open files

![image](https://github.com/microsoft/pylance-release/assets/1333179/6c43c6b4-0477-4bdf-bd8f-da13f8def071)

**Comment by memeplex** [comment]
> This is the intended behavior. If you open a file that is outside of your project, pylance applies the diagnostic rules that you've selected in your settings.

@erictraut  doesn't this contradict https://github.com/microsoft/pylance-release/issues/5452#issuecomment-1949212053 or am I misunderstanding you?
